File: calculator-csharp/CalculatorLib/CalculatorService/Lexer.cs
using System;
using System.Resources;
using System.Linq;
using System.Collections.Generic;
using static System.Globalization.NumberFormatInfo;

# pragma warning disable CA1307 // Specify StringComparison


namespace CalculatorService
{
    /// <summary>Lexical analyzer (tokenizer)</summary>
    public class Lexer
    {
        private static string WHITESPACE = " \n\t";
        private static string NUMBER = TokenType.NUMBER.Value;
        private static string SYMBOLS = TokenType.GetSymbols();

        private CharEnumerator chars;
        private char? currentChar;

        public Lexer(string text)
        {
            text = text ?? "";

            chars = text.GetEnumerator();
            advance();
        }

        private void advance()
        {
            chars.MoveNext();
            try { currentChar = chars.Current; }
            catch (InvalidOperationException) { currentChar = null; }
        }

        public IEnumerable<Token> GenerateTokens()
        {
            while (currentChar != null)
            {
                if (WHITESPACE.Contains(chars.Current))
                    advance();
                else if (NUMBER.Contains(chars.Current))
                    yield return generateNumber();
                else if (SYMBOLS.Contains(chars.Current))
                    yield return generateSymbol();
                else
                    throw new Exception($"Illegal Char {currentChar}");
            }
        }

        private Token generateNumber()
        {
            string numberStr = "";

            while (currentChar != null && NUMBER.Contains(currentChar.ToString()))
            {
                numberStr += currentChar;
                advance();
            }

            if (numberStr.Count(x => x == '.') > 1)
                throw new Exception($"too many dot in {numberStr}");
            if (numberStr.EndsWith("."))
                numberStr += "0";

            double value = double.Parse(numberStr, provider: InvariantInfo);
            return new Token(TokenType.NUMBER, value);
        }

        private Token generateSymbol()
        {
            var tokenType = TokenType.Map[chars.Current];
            advance();
            return new Token(tokenType);
        }
    }
}

File: calculator-python/src/lexer.py
from tokens import Token, TokenType

WHITESPACE = ' \n\t'
NUMBER = TokenType.NUMBER.value
SYMBOLS = TokenType.symbols()


class Lexer:
    '''lexical analyzer (tokenizer)'''

    def __init__(self, _text):
        self.text = iter(_text)
        self._advance()

    def _advance(self):
        try:
            self.current_char = next(self.text)
        except StopIteration:
            self.current_char = None

    def generate_tokens(self):
        while self.current_char:
            if self.current_char in WHITESPACE:
                self._advance()
            elif self.current_char in NUMBER:
                yield self._generate_number()
            elif self.current_char in SYMBOLS:
                yield self._generate_symbol()
            else:
                raise Exception(f"Illegal Char {self.current_char}")

    def _generate_number(self):
        number_str = ""

        while self.current_char:
            if self.current_char not in NUMBER:
                break

            number_str += self.current_char
            self._advance()

        if number_str.count(".") > 1:
            raise Exception("too many dot")
        if number_str.startswith('.'):
            number_str = '0' + number_str
        if number_str.endswith('.'):
            number_str += '0'

        return Token(TokenType.NUMBER, float(number_str))

    def _generate_symbol(self):
        token_type = TokenType(self.current_char)
        self._advance()

        return Token(token_type)
